File: laravel/app/Services/AuditService.php
<?php

namespace App\Services;

use PhpAmqpLib\Connection\AMQPStreamConnection;
use PhpAmqpLib\Message\AMQPMessage;
use Exception;
use Illuminate\Support\Facades\Log;

class AuditService
{
    private $connection;
    private $channel;
    private $serviceName = 'LaravelService';

    public function __construct()
    {
        try {
            $this->connection = new AMQPStreamConnection(
                config('rabbitmq.host'),
                config('rabbitmq.port'),
                config('rabbitmq.user'),
                config('rabbitmq.password')
            );

            $this->channel = $this->connection->channel();

            $this->channel->exchange_declare('audit_logs', 'topic', false, true, false);
        } catch (Exception $e) {
            Log::error('Failed to initialize RabbitMQ connection: ' . $e->getMessage());
        }
    }

    public function logAction(string $action, string $entity, string $description): void
    {
        try {
            if (!$this->channel) {
                Log::error('RabbitMQ channel not available');
                return;
            }

            $auditMessage = [
                'timestamp' => now()->toISOString(),
                'service' => $this->serviceName,
                'action' => $action,
                'entity' => $entity,
                'description' => $description
            ];

            $message = new AMQPMessage(
                json_encode($auditMessage),
                ['content_type' => 'application/json']
            );

            $this->channel->basic_publish($message, 'audit_logs', 'audit.laravelservice');

            Log::debug("Audit log sent: {$action} - {$entity} - {$description}");
        } catch (Exception $e) {
            Log::error('Error publishing audit log: ' . $e->getMessage());
        }
    }

    public function __destruct()
    {
        try {
            if ($this->channel) {
                $this->channel->close();
            }
            if ($this->connection) {
                $this->connection->close();
            }
        } catch (Exception $e) {
            Log::error('Error closing RabbitMQ connection: ' . $e->getMessage());
        }
    }
}

File: JavaService/src/main/java/org/student/microserviceapp/javaservice/services/AuditService.java
package org.student.microserviceapp.javaservice.services;

import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.extern.slf4j.Slf4j;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.util.HashMap;
import java.util.Map;

@Service
@Slf4j
public class AuditService {

    private final RabbitTemplate rabbitTemplate;
    private final ObjectMapper objectMapper;

    @Value("${spring.application.name:JavaService}")
    private String serviceName;

    public AuditService(RabbitTemplate rabbitTemplate, ObjectMapper objectMapper) {
        this.rabbitTemplate = rabbitTemplate;
        this.objectMapper = objectMapper;
    }

    public void logAction(String action, String entity, String description) {
        try {
            Map<String, Object> auditMessage = new HashMap<>();
            auditMessage.put("timestamp", LocalDateTime.now(ZoneOffset.UTC).toString());
            auditMessage.put("service", serviceName);
            auditMessage.put("action", action);
            auditMessage.put("entity", entity);
            auditMessage.put("description", description);

            rabbitTemplate.convertAndSend("audit_logs", "audit.javaservice", auditMessage);
            
            log.debug("Audit log sent: {} - {} - {}", action, entity, description);
        } catch (Exception ex) {
            log.error("Error publishing audit log: {}", ex.getMessage());
        }
    }
}